File: sdk/rust/zhtw/src/matcher.rs
//! Core matching engine: daachorse wrapper, identity protection, byte↔codepoint mapping.

use std::collections::{BTreeSet, HashSet};

use daachorse::CharwiseDoubleArrayAhoCorasick;

use crate::header;

// ── Types ────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub(crate) struct TermHit {
    pub byte_start: usize,
    pub byte_end: usize,
    pub source: String,
    pub target: String,
}

// ── Pattern table deserialization ────────────────────────────────────────────

/// Reads the binary pattern table format written by build.rs.
///
/// Format:
///   u32 LE: count
///   For each entry:
///     u8: source_mask (0b01=CN, 0b10=HK, 0b11=both)
///     u32 LE: source_len
///     [u8; source_len]: source bytes
///     u32 LE: target_len
///     [u8; target_len]: target bytes
///
/// Returns `(source, target, source_mask)` triples.
pub(crate) fn deserialize_pattern_table(bytes: &[u8]) -> Vec<(String, String, u8)> {
    let mut pos = 0;

    let count = u32::from_le_bytes(bytes[pos..pos + 4].try_into().unwrap()) as usize;
    pos += 4;

    let mut table = Vec::with_capacity(count);
    for _ in 0..count {
        let mask = bytes[pos];
        pos += 1;

        let src_len = u32::from_le_bytes(bytes[pos..pos + 4].try_into().unwrap()) as usize;
        pos += 4;
        let src = std::str::from_utf8(&bytes[pos..pos + src_len])
            .unwrap()
            .to_string();
        pos += src_len;

        let tgt_len = u32::from_le_bytes(bytes[pos..pos + 4].try_into().unwrap()) as usize;
        pos += 4;
        let tgt = std::str::from_utf8(&bytes[pos..pos + tgt_len])
            .unwrap()
            .to_string();
        pos += tgt_len;

        table.push((src, tgt, mask));
    }

    table
}

// ── Automaton construction ──────────────────────────────────────────────────

/// Build a runtime automaton from a pattern list (for custom dictionaries).
pub(crate) fn build_automaton(
    patterns: &[(String, String)],
) -> CharwiseDoubleArrayAhoCorasick<u32> {
    use daachorse::{CharwiseDoubleArrayAhoCorasickBuilder, MatchKind};

    let patvals: Vec<(&str, u32)> = patterns
        .iter()
        .enumerate()
        .map(|(i, (src, _))| (src.as_str(), i as u32))
        .collect();

    CharwiseDoubleArrayAhoCorasickBuilder::new()
        .match_kind(MatchKind::Standard)
        .build_with_values(patvals)
        .expect("failed to build daachorse automaton")
}

/// Deserialize the precompiled default automaton from embedded bytes.
///
/// # Safety
/// Uses `deserialize_unchecked` — safe when bytes come from the same daachorse
/// version (verified by header magic + version check).
pub(crate) fn deserialize_default_automaton(bytes: &[u8]) -> CharwiseDoubleArrayAhoCorasick<u32> {
    let payload = header::verify_header(bytes, header::SourceMask::CN_HK);
    // SAFETY: bytes were produced by the same daachorse version (verified above).
    let (automaton, _trailing) =
        unsafe { CharwiseDoubleArrayAhoCorasick::deserialize_unchecked(payload) };
    automaton
}

// ── Byte ↔ codepoint mapping ────────────────────────────────────────────────

/// Build a mapping from byte offset to codepoint index.
///
/// Returns a `Vec` of length `text.len() + 1` where `map[byte_offset]` gives
/// the codepoint index at that byte position.
pub(crate) fn build_byte_to_cp(text: &str) -> Vec<usize> {
    let mut map = vec![0usize; text.len() + 1];
    for (cp_idx, (byte_idx, ch)) in text.char_indices().enumerate() {
        map[byte_idx] = cp_idx;
        // Fill intermediate bytes within this char with the same cp index.
        for item in map
            .iter_mut()
            .take(byte_idx + ch.len_utf8())
            .skip(byte_idx + 1)
        {
            *item = cp_idx;
        }
    }
    // Sentinel: one past the last byte = total codepoint count.
    map[text.len()] = text.chars().count();
    map
}

// ── Core matching algorithm ─────────────────────────────────────────────────

/// Find all non-overlapping term matches using the Aho-Corasick automaton,
/// with identity-based protection to avoid false conversions.
///
/// This is the Rust port of Python `matcher.py:find_matches` (lines 60-134).
pub(crate) fn find_term_matches(
    pma: &CharwiseDoubleArrayAhoCorasick<u32>,
    pattern_table: &[(String, String)],
    text: &str,
) -> Vec<TermHit> {
    // 1. Collect all overlapping hits.
    let mut all_hits: Vec<TermHit> = pma
        .find_overlapping_iter(text)
        .map(|m| {
            let idx = m.value() as usize;
            let (ref src, ref tgt) = pattern_table[idx];
            TermHit {
                byte_start: m.start(),
                byte_end: m.end(),
                source: src.clone(),
                target: tgt.clone(),
            }
        })
        .collect();

    if all_hits.is_empty() {
        return Vec::new();
    }

    // 2. Sort by (byte_start ASC, length DESC).
    all_hits.sort_by(|a, b| {
        a.byte_start.cmp(&b.byte_start).then_with(|| {
            let len_a = a.byte_end - a.byte_start;
            let len_b = b.byte_end - b.byte_start;
            len_b.cmp(&len_a) // DESC
        })
    });

    // 3. Build protected byte-range set from identity matches.
    let mut identity_matches: Vec<&TermHit> = Vec::new();
    let mut non_identity_spans: Vec<(usize, usize)> = Vec::new();

    for hit in &all_hits {
        if hit.source == hit.target {
            identity_matches.push(hit);
        } else {
            non_identity_spans.push((hit.byte_start, hit.byte_end));
        }
    }

    let mut protected: BTreeSet<usize> = BTreeSet::new();

    if non_identity_spans.is_empty() {
        // All identity → all ranges are protected (nothing to convert).
        for hit in &identity_matches {
            for b in hit.byte_start..hit.byte_end {
                protected.insert(b);
            }
        }
    } else {
        // Sort non-identity spans by start, build prefix_max_end.
        non_identity_spans.sort_by_key(|&(s, _)| s);

        let mut prefix_max_end: Vec<usize> = Vec::with_capacity(non_identity_spans.len());
        let mut running_max = 0usize;
        for &(_, end) in &non_identity_spans {
            running_max = running_max.max(end);
            prefix_max_end.push(running_max);
        }

        // For each identity match, check if it's fully contained in any non-identity span.
        for hit in &identity_matches {
            if !is_contained_in_non_identity(
                hit.byte_start,
                hit.byte_end,
                &non_identity_spans,
                &prefix_max_end,
            ) {
                for b in hit.byte_start..hit.byte_end {
                    protected.insert(b);
                }
            }
        }
    }

    // 4. Left-to-right greedy filter.
    let mut result: Vec<TermHit> = Vec::new();
    let mut cursor: usize = 0;

    for hit in &all_hits {
        if hit.byte_start < cursor {
            continue; // overlap — skip
        }

        let is_identity = hit.source == hit.target;

        if !is_identity {
            // Check if any byte in [start..end) is protected.
            if (hit.byte_start..hit.byte_end).any(|b| protected.contains(&b)) {
                continue; // protected — skip (don't advance cursor)
            }
        }

        cursor = hit.byte_end;

        if !is_identity {
            result.push(hit.clone());
        }
        // Identity matches advance cursor but are never emitted.
    }

    result
}

/// Return all byte positions covered by any raw automaton hit, including
/// identity terms (source == target). Used to prevent the char layer from
/// converting characters protected by identity term matches.
pub(crate) fn get_covered_positions(
    pma: &CharwiseDoubleArrayAhoCorasick<u32>,
    text: &str,
) -> HashSet<usize> {
    let mut covered = HashSet::new();
    for m in pma.find_overlapping_iter(text) {
        for b in m.start()..m.end() {
            covered.insert(b);
        }
    }
    covered
}

/// Check whether the identity span `[start, end)` is fully contained within
/// some non-identity span, using binary search + prefix_max_end.
fn is_contained_in_non_identity(
    start: usize,
    end: usize,
    spans: &[(usize, usize)],
    prefix_max_end: &[usize],
) -> bool {
    // Find the rightmost span with span.start <= start.
    let idx = match spans.binary_search_by_key(&start, |&(s, _)| s) {
        Ok(i) => i,
        Err(0) => return false, // All spans start after this identity match.
        Err(i) => i - 1,
    };

    // Check if the prefix_max_end up to idx covers end.
    // Any span in [0..=idx] that starts <= start and ends >= end would contain it.
    prefix_max_end[idx] >= end
}

// ── Charmap application ────────────────────────────────────────────────────

/// Apply charmap to a text segment, skipping byte positions that are in the
/// covered set. `offset` is the byte offset of this segment within the
/// original text.
pub(crate) fn apply_charmap_skipping(
    segment: &str,
    char_map: &phf::Map<char, char>,
    covered: &HashSet<usize>,
    offset: usize,
) -> String {
    let mut result = String::with_capacity(segment.len());
    for (byte_idx, ch) in segment.char_indices() {
        if covered.contains(&(offset + byte_idx)) {
            result.push(ch);
        } else {
            result.push(char_map.get(&ch).copied().unwrap_or(ch));
        }
    }
    result
}

// ── Tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn byte_to_cp_ascii() {
        let map = build_byte_to_cp("hello");
        assert_eq!(map[0], 0);
        assert_eq!(map[1], 1);
        assert_eq!(map[5], 5);
    }

    #[test]
    fn byte_to_cp_cjk() {
        // "中X" — 中=3 bytes, X=1 byte
        let map = build_byte_to_cp("中X");
        assert_eq!(map[0], 0);
        assert_eq!(map[3], 1);
        assert_eq!(map[4], 2);
    }

    #[test]
    fn byte_to_cp_supplementary() {
        // "𠮷A" — 𠮷=4 bytes, A=1 byte
        let map = build_byte_to_cp("𠮷A");
        assert_eq!(map[0], 0);
        assert_eq!(map[4], 1);
        assert_eq!(map[5], 2);
    }
}

File: sdk/go/zhtw/matcher.go
package zhtw

import "sort"

// ── Internal types ──────────────────────────────────────────────────────────

type acPattern struct {
	source  string
	target  string
	runeLen int // len([]rune(source)), precomputed
}

type acMatch struct {
	start  int // rune index, inclusive
	end    int // rune index, exclusive
	source string
	target string
}

type matchDecision struct {
	hit        acMatch
	outcome    string
	reasonCode string
}

type acNode struct {
	children map[rune]*acNode
	fail     *acNode
	outputs  []int // indices into ahoCorasick.patterns
}

type ahoCorasick struct {
	root     *acNode
	patterns []acPattern
}

// ── Construction ────────────────────────────────────────────────────────────

func newACNode() *acNode {
	return &acNode{children: make(map[rune]*acNode)}
}

func buildAhoCorasick(patterns []acPattern) *ahoCorasick {
	ac := &ahoCorasick{
		root:     newACNode(),
		patterns: patterns,
	}
	// Insert patterns into trie.
	for idx, pat := range patterns {
		node := ac.root
		for _, r := range pat.source {
			next, ok := node.children[r]
			if !ok {
				next = newACNode()
				node.children[r] = next
			}
			node = next
		}
		node.outputs = append(node.outputs, idx)
	}
	// Build failure links (BFS).
	queue := make([]*acNode, 0)
	for _, child := range ac.root.children {
		child.fail = ac.root
		queue = append(queue, child)
	}
	for head := 0; head < len(queue); head++ {
		node := queue[head]
		for r, child := range node.children {
			queue = append(queue, child)
			f := node.fail
			for f != nil && f.children[r] == nil {
				f = f.fail
			}
			if f == nil {
				child.fail = ac.root
			} else if next, ok := f.children[r]; ok {
				child.fail = next
			} else {
				child.fail = ac.root
			}
			// Merge output-link patterns from fail node.
			if len(child.fail.outputs) > 0 {
				merged := make([]int, len(child.outputs)+len(child.fail.outputs))
				copy(merged, child.outputs)
				copy(merged[len(child.outputs):], child.fail.outputs)
				child.outputs = merged
			}
		}
	}
	return ac
}

// ── Raw emissions ───────────────────────────────────────────────────────────

// iterEmissions returns ALL overlapping AC hits (rune-indexed).
func (ac *ahoCorasick) iterEmissions(runes []rune) []acMatch {
	var hits []acMatch
	node := ac.root
	for i, r := range runes {
		for node != ac.root && node.children[r] == nil {
			node = node.fail
		}
		if next, ok := node.children[r]; ok {
			node = next
		}
		for _, idx := range node.outputs {
			pat := ac.patterns[idx]
			start := i + 1 - pat.runeLen
			hits = append(hits, acMatch{
				start:  start,
				end:    i + 1,
				source: pat.source,
				target: pat.target,
			})
		}
	}
	return hits
}

// scan walks the automaton once and derives selected matches and effective coverage.
func (ac *ahoCorasick) scan(runes []rune) ([]acMatch, map[int]bool) {
	raw := ac.iterEmissions(runes)
	matches, covered := selectTermMatches(raw)
	for _, m := range matches {
		for i := m.start; i < m.end; i++ {
			covered[i] = true
		}
	}
	return matches, covered
}

func (ac *ahoCorasick) scanDetailed(runes []rune) ([]acMatch, map[int]bool, []matchDecision) {
	matches, covered, decisions := selectTermMatchesCore(ac.iterEmissions(runes), true)
	for _, match := range matches {
		for i := match.start; i < match.end; i++ {
			covered[i] = true
		}
	}
	return matches, covered, decisions
}

// ── Match selection ──────────────────────────────────────────────────────────

// findTermMatches returns non-overlapping, leftmost-longest term matches
// with identity-based protection. Identity terms (source == target) advance
// the cursor but are never emitted. Non-identity terms overlapping a
// protected position are skipped without advancing the cursor.
//
// Mirrors Python src/zhtw/matcher.py:find_matches.
func (ac *ahoCorasick) findTermMatches(runes []rune) []acMatch {
	matches, _ := selectTermMatches(ac.iterEmissions(runes))
	return matches
}

func selectTermMatches(raw []acMatch) ([]acMatch, map[int]bool) {
	matches, covered, _ := selectTermMatchesCore(raw, false)
	return matches, covered
}

func selectTermMatchesCore(raw []acMatch, detailed bool) ([]acMatch, map[int]bool, []matchDecision) {
	if len(raw) == 0 {
		return nil, make(map[int]bool), nil
	}

	// Sort by (start ASC, length DESC).
	sort.Slice(raw, func(i, j int) bool {
		if raw[i].start != raw[j].start {
			return raw[i].start < raw[j].start
		}
		lenI := raw[i].end - raw[i].start
		lenJ := raw[j].end - raw[j].start
		return lenI > lenJ
	})

	// Separate identity and non-identity hits.
	var identity []acMatch
	var nonIdentitySpans [][2]int
	for _, m := range raw {
		if m.source == m.target {
			identity = append(identity, m)
		} else {
			nonIdentitySpans = append(nonIdentitySpans, [2]int{m.start, m.end})
		}
	}

	// Build protected rune positions from identity matches not contained
	// in any non-identity span.
	protected := make(map[int]bool)
	effectiveIdentity := make(map[acMatch]bool)
	if len(nonIdentitySpans) == 0 {
		for _, m := range identity {
			for i := m.start; i < m.end; i++ {
				protected[i] = true
			}
			effectiveIdentity[m] = true
		}
	} else {
		sort.Slice(nonIdentitySpans, func(i, j int) bool {
			return nonIdentitySpans[i][0] < nonIdentitySpans[j][0]
		})
		// Build prefix-max-end array.
		prefixMaxEnd := make([]int, len(nonIdentitySpans))
		maxEnd := 0
		for i, span := range nonIdentitySpans {
			if span[1] > maxEnd {
				maxEnd = span[1]
			}
			prefixMaxEnd[i] = maxEnd
		}
		niStarts := make([]int, len(nonIdentitySpans))
		for i, span := range nonIdentitySpans {
			niStarts[i] = span[0]
		}
		for _, m := range identity {
			idx := bisectRight(niStarts, m.start) - 1
			contained := idx >= 0 && prefixMaxEnd[idx] >= m.end
			if !contained {
				for i := m.start; i < m.end; i++ {
					protected[i] = true
				}
				effectiveIdentity[m] = true
			}
		}
	}

	// Left-to-right greedy filter.
	var result []acMatch
	var decisions []matchDecision
	cursor := 0
	for _, m := range raw {
		if m.start < cursor {
			if detailed {
				isIdentity := m.source == m.target
				effective := effectiveIdentity[m]
				outcome := "skipped"
				reason := "overlap_loser"
				if isIdentity {
					if effective {
						outcome = "protected"
						reason = "identity_guard"
					} else {
						reason = "identity_contained"
					}
				}
				decisions = append(decisions, matchDecision{hit: m, outcome: outcome, reasonCode: reason})
			}
			continue
		}
		isIdentity := m.source == m.target
		if !isIdentity {
			overlaps := false
			for i := m.start; i < m.end; i++ {
				if protected[i] {
					overlaps = true
					break
				}
			}
			if overlaps {
				if detailed {
					decisions = append(decisions, matchDecision{
						hit: m, outcome: "skipped", reasonCode: "protected_by_identity",
					})
				}
				continue // skip without advancing cursor
			}
		}
		cursor = m.end
		if !isIdentity {
			result = append(result, m)
			if detailed {
				decisions = append(decisions, matchDecision{
					hit: m, outcome: "applied", reasonCode: "term_selected",
				})
			}
		} else if detailed {
			reason := "identity_contained"
			if effectiveIdentity[m] {
				reason = "identity_guard"
			}
			decisions = append(decisions, matchDecision{
				hit: m, outcome: "protected", reasonCode: reason,
			})
		}
	}
	return result, protected, decisions
}

// bisectRight returns the insertion point for x in a sorted slice a
// such that all elements before the point are <= x.
func bisectRight(a []int, x int) int {
	return sort.Search(len(a), func(i int) bool { return a[i] > x })
}
